Composite rules of x3::string("..."), if assigned an action, cannot be used in phrase_parse.

This is really weird and I cannot explain what's happening here. Possibly I'm stupid and not seeing something obvious, but I don't think that's the case. Here's an example of what I'm trying to do:

```cpp
    namespace x3 = boost::spirit::x3;
    auto print_attr_type = [=](auto& ctx) {
        std::cout << typeid (x3::_attr(ctx)).name() << std::endl;
    };

    std::string s("\"Text\"");
    auto parser1 = x3::lexeme['"' >> *(x3::char_('\\') >> x3::char_('"') | x3::char_ - '"') >> '"'][print_attr_type];
    auto parser2 = x3::lexeme['"' >> *(x3::string("\\\"") | x3::char_ - '"') >> '"'][print_attr_type];
    x3::phrase_parse(s.begin(),
                     s.end(),
                     parser1,
                     x3::space);
```

Now, this compiles all fine, but if I change the parser in the parsing call to parser2, I get the weirdest compiler errors. The one where it seems to break down is

```
/usr/include/boost/spirit/home/x3/support/traits/move_to.hpp:201:24: error: no matching function for call to ‘move_to(std::remove_reference<char&>::type, std::vector<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<2>, std::__cxx11::basic_string<char>, boost::mpl::l_item<mpl_::long_<1>, char, boost::mpl::l_end> > > >, std::allocator<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<2>, std::__cxx11::basic_string<char>, boost::mpl::l_item<mpl_::long_<1>, char, boost::mpl::l_end> > > > > >&, boost::mpl::identity<boost::spirit::x3::traits::container_attribute>::type)’
  201 |         detail::move_to(std::move(src), dest
      |         ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
  202 |           , typename attribute_category<Dest>::type());
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Assigning an action to `x3::string("\\\"")` alone works absolutely fine, and the definition of rule2 itself also seems to work fine, but using rule2 fails. I would be very happy if someone could help me out here, since rule2 looks nicer in code than rule1 (and should, in theory, result in a std::string attribute instead of a `deque<char,char>` one).

--- a/include/boost/spirit/home/x3/operator/detail/alternative.hpp
+++ b/include/boost/spirit/home/x3/operator/detail/alternative.hpp
@@ -258,6 +258,22 @@ namespace boost { namespace spirit { namespace x3 { namespace detail
         return false;
     }
 
+    template <typename Subject>
+    struct alternative_helper : unary_parser<Subject, alternative_helper<Subject>>
+    {
+        static bool const is_pass_through_unary = true;
+
+        using unary_parser<Subject, alternative_helper<Subject>>::unary_parser;
+
+        template <typename Iterator, typename Context
+          , typename RContext, typename Attribute>
+        bool parse(Iterator& first, Iterator const& last
+          , Context const& context, RContext& rcontext, Attribute& attr) const
+        {
+            return detail::parse_alternative(this->subject, first, last, context, rcontext, attr);
+        }
+    };
+
     template <typename Left, typename Right, typename Context, typename RContext>
     struct parse_into_container_impl<alternative<Left, Right>, Context, RContext>
     {
@@ -267,33 +283,10 @@ namespace boost { namespace spirit { namespace x3 { namespace detail
         static bool call(
             parser_type const& parser
           , Iterator& first, Iterator const& last
-          , Context const& context, RContext& rcontext, Attribute& attr, mpl::true_)
-        {
-            return parse_alternative(parser, first, last, context, rcontext, attr);
-        }
-
-        template <typename Iterator, typename Attribute>
-        static bool call(
-            parser_type const& parser
-          , Iterator& first, Iterator const& last
-          , Context const& context, RContext& rcontext, Attribute& attr, mpl::false_)
-        {
-            return parse_into_container_base_impl<parser_type>::call(
-                parser, first, last, context, rcontext, attr);
-        }
-
-        template <typename Iterator, typename Attribute>
-        static bool call(
-            parser_type const& parser
-          , Iterator& first, Iterator const& last
           , Context const& context, RContext& rcontext, Attribute& attr)
         {
-            typedef typename
-                traits::attribute_of<parser_type, Context>::type
-            attribute_type;
-
-            return call(parser, first, last, context, rcontext, attr
-                , traits::variant_has_substitute<attribute_type, Attribute>());
+            return detail::parse_into_container(alternative_helper<Left>{parser.left}, first, last, context, rcontext, attr)
+                || detail::parse_into_container(alternative_helper<Right>{parser.right}, first, last, context, rcontext, attr);
         }
     };
 

--- a/test/x3/alternative.cpp
+++ b/test/x3/alternative.cpp
@@ -250,5 +250,25 @@ main()
         BOOST_TEST_EQ(st.val, 42);
     }
 
+    { // attributeless parsers must not insert values
+        std::vector<int> v;
+        BOOST_TEST(test_attr("1 2 3 - 5 - - 7 -", (int_ | '-') % ' ', v));
+        BOOST_TEST_EQ(v.size(), 5)
+            && BOOST_TEST_EQ(v[0], 1)
+            && BOOST_TEST_EQ(v[1], 2)
+            && BOOST_TEST_EQ(v[2], 3)
+            && BOOST_TEST_EQ(v[3], 5)
+            && BOOST_TEST_EQ(v[4], 7)
+            ;
+    }
+
+    { // regressing test for #603
+        using boost::spirit::x3::attr;
+        struct X {};
+        std::vector<boost::variant<std::string, int, X>> v;
+        BOOST_TEST(test_attr("xx42x9y", *(int_ | +char_('x') | 'y' >> attr(X{})), v));
+        BOOST_TEST_EQ(v.size(), 5);
+    }
+
     return boost::report_errors();
 }

